Sources/NetworkProtection/Diagnostics/NetworkProtectionLatencyMonitor.swift

```diff
@@ -0,0 +1,261 @@
+        } catch {
+            os_log("⚫️ Stopping latency monitor prematurely", log: log)
+            throw error
+        }
+    }
+
+    func stop() async {
+        os_log("⚫️ Stopping latency monitor", log: log)
+        await stopScheduledTimer()
+    }
+
+    // MARK: - Timer scheduling
+
+    private func scheduleTimer() async throws {
+        await stopScheduledTimer()
+
+        await timerRunCoordinator.start()
+
+        let timer = DispatchSource.makeTimerSource(queue: timerQueue)
+        self.timer = timer
+
+        timer.schedule(deadline: .now() + Self.measurementInterval, repeating: Self.measurementInterval)
+        timer.setEventHandler { [weak self] in
+            guard let self else { return }
+
+            Task {
+                await self.measureLatency()
+            }
+        }
+
+        timer.setCancelHandler { [weak self] in
+            self?.timer = nil
+        }
+
+        timer.resume()
+    }
+
+    private func stopScheduledTimer() async {
+        await timerRunCoordinator.stop()
+
+        cancelTimerImmediately()
+    }
+
+    private func cancelTimerImmediately() {
+        guard let timer else { return }
+
+        if !timer.isCancelled {
+            timer.cancel()
+        }
+
+        self.timer = nil
+    }
+
+    // MARK: - Latency monitor
+
+    @MainActor
+    func measureLatency() async {
+        guard let serverIP = serverIP() else {
+            latencySubject.send(Self.unknownLatency)
+            return
+        }
+
+        os_log("⚫️ Pinging %{public}s", log: .networkProtectionPixel, type: .debug, serverIP.debugDescription)
+
+        let result = await Pinger(ip: serverIP, timeout: Self.pingTimeout, log: .networkProtectionPixel).ping()
+
+        switch result {
+        case .success(let pingResult):
+            latencySubject.send(pingResult.time)
+        case .failure(let error):
+            os_log("⚫️ Ping error: %{public}s", log: .networkProtectionPixel, type: .debug, error.localizedDescription)
+            latencySubject.send(Self.unknownLatency)
+        }
+    }
+}
+
+final class ExponentialGeometricAverage {
+    private static let decayConstant = 0.1
+    private let cutover = ceil(1 / decayConstant)
+
+    private var count = TimeInterval(0)
+    private var value = TimeInterval(-1)
+
+    var average: TimeInterval {
+        value
+    }
+
+    func addMeasurement(_ measurement: TimeInterval) {
+        let keepConstant = 1 - Self.decayConstant
+
+        if count > cutover {
+            value = exp(keepConstant * log(value) + Self.decayConstant * log(measurement))
+        } else if count > 0 {
+            let retained: Double = keepConstant * count / (count + 1.0)
+            let newcomer = 1.0 - retained
+            value = exp(retained * log(value) + newcomer * log(measurement))
+        } else {
+            value = measurement
+        }
+        count += 1
+    }
+
+    func reset() {
+        value = -1.0
+        count = 0
+    }
+}
```

Comment:
Straight port of https://github.com/duckduckgo/Android/blob/develop/network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/connectionclass/ExponentialGeometricAverage.kt

Reply:
Looks good - what do you think about adding some unit tests to it?